File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/zk/VerifyingKeyRecordDescription.java
package org.hyperledger.iroha.android.model.zk;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;

/**
 * Describes the fields embedded in {@code RegisterVerifyingKey} and
 * {@code UpdateVerifyingKey} instructions.
 *
 * <p>The builder enforces the same invariants as the server-side DTO handling:
 * either inline verifying key bytes must be provided or the commitment + length
 * pair must be supplied. Gas schedule identifiers are required, and the builder
 * defaults optional fields to sensible values when omitted.
 */
public final class VerifyingKeyRecordDescription {

  private final int version;
  private final String circuitId;
  private final VerifyingKeyBackendTag backendTag;
  private final String curve;
  private final String schemaHashHex;
  private final String commitmentHex;
  private final byte[] inlineKeyBytes;
  private final Integer vkLength;
  private final Integer maxProofBytes;
  private final String gasScheduleId;
  private final String metadataUriCid;
  private final String vkBytesCid;
  private final Long activationHeight;
  private final Long withdrawHeight;
  private final VerifyingKeyStatus status;

  private VerifyingKeyRecordDescription(final Builder builder, final String computedCommitmentHex) {
    this.version = builder.version;
    this.circuitId = builder.circuitId;
    this.backendTag = builder.backendTag;
    this.curve = builder.curve;
    this.schemaHashHex = builder.schemaHashHex;
    this.commitmentHex = computedCommitmentHex;
    this.inlineKeyBytes = builder.inlineKeyBytes;
    this.vkLength = builder.vkLength;
    this.maxProofBytes = builder.maxProofBytes;
    this.gasScheduleId = builder.gasScheduleId;
    this.metadataUriCid = builder.metadataUriCid;
    this.vkBytesCid = builder.vkBytesCid;
    this.activationHeight = builder.activationHeight;
    this.withdrawHeight = builder.withdrawHeight;
    this.status = builder.status;
  }

  public int version() {
    return version;
  }

  public String circuitId() {
    return circuitId;
  }

  public VerifyingKeyBackendTag backendTag() {
    return backendTag;
  }

  public String schemaHashHex() {
    return schemaHashHex;
  }

  public String commitmentHex() {
    return commitmentHex;
  }

  public byte[] inlineKeyBytes() {
    return inlineKeyBytes == null ? null : inlineKeyBytes.clone();
  }

  public Integer vkLength() {
    return vkLength;
  }

  public Integer maxProofBytes() {
    return maxProofBytes;
  }

  public String gasScheduleId() {
    return gasScheduleId;
  }

  public VerifyingKeyStatus status() {
    return status;
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof VerifyingKeyRecordDescription other)) {
      return false;
    }
    return version == other.version
        && Objects.equals(circuitId, other.circuitId)
        && backendTag == other.backendTag
        && Objects.equals(curve, other.curve)
        && Objects.equals(schemaHashHex, other.schemaHashHex)
        && Objects.equals(commitmentHex, other.commitmentHex)
        && Arrays.equals(inlineKeyBytes, other.inlineKeyBytes)
        && Objects.equals(vkLength, other.vkLength)
        && Objects.equals(maxProofBytes, other.maxProofBytes)
        && Objects.equals(gasScheduleId, other.gasScheduleId)
        && Objects.equals(metadataUriCid, other.metadataUriCid)
        && Objects.equals(vkBytesCid, other.vkBytesCid)
        && Objects.equals(activationHeight, other.activationHeight)
        && Objects.equals(withdrawHeight, other.withdrawHeight)
        && status == other.status;
  }

  @Override
  public int hashCode() {
    int result =
        Objects.hash(
            version,
            circuitId,
            backendTag,
            curve,
            schemaHashHex,
            commitmentHex,
            vkLength,
            maxProofBytes,
            gasScheduleId,
            metadataUriCid,
            vkBytesCid,
            activationHeight,
            withdrawHeight,
            status);
    result = 31 * result + Arrays.hashCode(inlineKeyBytes);
    return result;
  }

  /**
    * Serialises the record into Norito-style arguments using the provided backend identifier for
    * inline verifying key bytes.
    */
  public Map<String, String> toArguments(final String backend) {
    Objects.requireNonNull(backend, "backend");
    final Map<String, String> args = new LinkedHashMap<>();
    args.put("record.version", Integer.toUnsignedString(version));
    args.put("record.circuit_id", circuitId);
    args.put("record.backend_tag", backendTag.noritoValue());
    args.put("record.curve", curve);
    args.put("record.public_inputs_schema_hash_hex", schemaHashHex);
    args.put("record.commitment_hex", commitmentHex);
    if (inlineKeyBytes != null) {
      args.put(
          "record.vk_bytes_b64",
          Base64.getEncoder().encodeToString(inlineKeyBytes));
      args.put("record.vk_len", Integer.toUnsignedString(vkLength));
    } else if (vkLength != null) {
      args.put("record.vk_len", Integer.toUnsignedString(vkLength));
    }
    if (maxProofBytes != null) {
      args.put("record.max_proof_bytes", Integer.toUnsignedString(maxProofBytes));
    }
    args.put("record.gas_schedule_id", gasScheduleId);
    if (metadataUriCid != null) {
      args.put("record.metadata_uri_cid", metadataUriCid);
    }
    if (vkBytesCid != null) {
      args.put("record.vk_bytes_cid", vkBytesCid);
    }
    if (activationHeight != null) {
      args.put("record.activation_height", Long.toUnsignedString(activationHeight));
    }
    if (withdrawHeight != null) {
      args.put("record.withdraw_height", Long.toUnsignedString(withdrawHeight));
    }
    args.put("record.status", status.wireName());
    return args;
  }

  public static Builder builder() {
    return new Builder();
  }

  public static final class Builder {
    private Integer version;
    private String circuitId;
    private VerifyingKeyBackendTag backendTag = VerifyingKeyBackendTag.UNSUPPORTED;
    private String curve = "unknown";
    private String schemaHashHex;
    private String commitmentHex;
    private byte[] inlineKeyBytes;
    private Integer vkLength;
    private Integer maxProofBytes;
    private String gasScheduleId;
    private String metadataUriCid;
    private String vkBytesCid;
    private Long activationHeight;
    private Long withdrawHeight;
    private VerifyingKeyStatus status = VerifyingKeyStatus.ACTIVE;

    private Builder() {}

    public Builder setVersion(final int version) {
      if (version < 0) {
        throw new IllegalArgumentException("version must be non-negative");
      }
      this.version = version;
      return this;
    }

    public Builder setCircuitId(final String circuitId) {
      if (circuitId == null || circuitId.trim().isEmpty()) {
        throw new IllegalArgumentException("circuitId must not be blank");
      }
      this.circuitId = circuitId.trim();
      return this;
    }

    public Builder setBackendTag(final VerifyingKeyBackendTag backendTag) {
      this.backendTag = Objects.requireNonNull(backendTag, "backendTag");
      return this;
    }

    public Builder setCurve(final String curve) {
      if (curve != null && !curve.trim().isEmpty()) {
        this.curve = curve.trim();
      }
      return this;
    }

    public Builder setSchemaHashHex(final String schemaHashHex) {
      this.schemaHashHex = validateHex(schemaHashHex, "public_inputs_schema_hash_hex");
      return this;
    }

    public Builder setCommitmentHex(final String commitmentHex) {
      this.commitmentHex = validateHex(commitmentHex, "commitment_hex");
      return this;
    }

    public Builder setInlineKeyBytes(final byte[] inlineKeyBytes) {
      if (inlineKeyBytes == null) {
        this.inlineKeyBytes = null;
      } else {
        this.inlineKeyBytes = inlineKeyBytes.clone();
      }
      return this;
    }

    public Builder setVkLength(final Integer vkLength) {
      if (vkLength != null && vkLength <= 0) {
        throw new IllegalArgumentException("vkLength must be greater than zero");
      }
      this.vkLength = vkLength;
      return this;
    }

    public Builder setMaxProofBytes(final Integer maxProofBytes) {
      if (maxProofBytes != null && maxProofBytes < 0) {
        throw new IllegalArgumentException("maxProofBytes must be non-negative");
      }
      this.maxProofBytes = maxProofBytes;
      return this;
    }

    public Builder setGasScheduleId(final String gasScheduleId) {
      if (gasScheduleId == null || gasScheduleId.trim().isEmpty()) {
        throw new IllegalArgumentException("gasScheduleId must not be blank");
      }
      this.gasScheduleId = gasScheduleId.trim();
      return this;
    }

    public Builder setMetadataUriCid(final String metadataUriCid) {
      this.metadataUriCid = normalizeOptional(metadataUriCid);
      return this;
    }

    public Builder setVkBytesCid(final String vkBytesCid) {
      this.vkBytesCid = normalizeOptional(vkBytesCid);
      return this;
    }

    public Builder setActivationHeight(final Long activationHeight) {
      if (activationHeight != null && activationHeight < 0) {
        throw new IllegalArgumentException("activationHeight must be non-negative");
      }
      this.activationHeight = activationHeight;
      return this;
    }

    public Builder setWithdrawHeight(final Long withdrawHeight) {
      if (withdrawHeight != null && withdrawHeight < 0) {
        throw new IllegalArgumentException("withdrawHeight must be non-negative");
      }
      this.withdrawHeight = withdrawHeight;
      return this;
    }

    public Builder setStatus(final VerifyingKeyStatus status) {
      this.status = Objects.requireNonNull(status, "status");
      return this;
    }

    public VerifyingKeyRecordDescription build(final String backend) {
      Objects.requireNonNull(backend, "backend");
      if (version == null) {
        throw new IllegalStateException("version must be provided");
      }
      if (circuitId == null) {
        throw new IllegalStateException("circuitId must be provided");
      }
      if (schemaHashHex == null) {
        throw new IllegalStateException("public_inputs_schema_hash_hex must be provided");
      }
      if (gasScheduleId == null) {
        throw new IllegalStateException("gasScheduleId must be provided");
      }
      final byte[] inlineBytes = inlineKeyBytes == null ? null : inlineKeyBytes.clone();
      final Integer lengthValue;
      if (inlineBytes != null) {
        lengthValue = inlineBytes.length;
        if (vkLength != null && !vkLength.equals(lengthValue)) {
          throw new IllegalStateException("vkLength does not match length of inline key bytes");
        }
      } else {
        if (commitmentHex == null) {
          throw new IllegalStateException("commitmentHex must be provided when vk bytes are absent");
        }
        if (vkLength == null) {
          throw new IllegalStateException("vkLength must be provided when vk bytes are absent");
        }
        lengthValue = vkLength;
      }
      if (withdrawHeight != null && activationHeight != null
          && withdrawHeight < activationHeight) {
        throw new IllegalStateException("withdrawHeight must be >= activationHeight");
      }
      final String computedCommitment =
          inlineBytes != null
              ? computeCommitmentHex(backend, inlineBytes)
              : commitmentHex.toLowerCase(Locale.ROOT);
      if (commitmentHex != null
          && inlineBytes != null
          && !commitmentHex.equalsIgnoreCase(computedCommitment)) {
        throw new IllegalStateException("commitmentHex does not match computed inline key commitment");
      }
      this.vkLength = lengthValue;
      return new VerifyingKeyRecordDescription(this, computedCommitment);
    }

    private static String validateHex(final String value, final String field) {
      if (value == null) {
        return null;
      }
      final String normalized = value.trim();
      if (normalized.isEmpty()) {
        return null;
      }
      if (normalized.length() != 64) {
        throw new IllegalArgumentException(field + " must contain exactly 64 hexadecimal characters");
      }
      for (int index = 0; index < normalized.length(); index++) {
        final char c = normalized.charAt(index);
        if (!((c >= '0' && c <= '9')
            || (c >= 'a' && c <= 'f')
            || (c >= 'A' && c <= 'F'))) {
          throw new IllegalArgumentException(field + " must contain only hexadecimal characters");
        }
      }
      return normalized.toLowerCase(Locale.ROOT);
    }

    private static String normalizeOptional(final String value) {
      if (value == null) {
        return null;
      }
      final String trimmed = value.trim();
      return trimmed.isEmpty() ? null : trimmed;
    }

    private static String computeCommitmentHex(final String backend, final byte[] bytes) {
      try {
        final MessageDigest digest = MessageDigest.getInstance("SHA-256");
        digest.update(backend.getBytes());
        digest.update(bytes);
        final byte[] hash = digest.digest();
        final StringBuilder builder = new StringBuilder(hash.length * 2);
        for (final byte b : hash) {
          builder.append(String.format("%02x", b));
        }
        return builder.toString();
      } catch (final NoSuchAlgorithmException ex) {
        throw new IllegalStateException("SHA-256 algorithm unavailable", ex);
      }
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/zk/VerifyingKeyRecordDescription.kt
@file:OptIn(ExperimentalEncodingApi::class)

package org.hyperledger.iroha.sdk.core.model.zk

import java.security.MessageDigest
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

/**
 * Describes the fields embedded in `RegisterVerifyingKey` and
 * `UpdateVerifyingKey` instructions.
 *
 * The factory method enforces the same invariants as the server-side DTO handling:
 * either inline verifying key bytes must be provided or the commitment + length
 * pair must be supplied. Gas schedule identifiers are required, and
 * optional fields default to sensible values when omitted.
 */
class VerifyingKeyRecordDescription private constructor(
    val version: Int,
    val circuitId: String,
    val backendTag: VerifyingKeyBackendTag,
    val curve: String,
    val schemaHashHex: String,
    val commitmentHex: String,
    inlineKeyBytes: ByteArray?,
    val vkLength: Int?,
    val maxProofBytes: Int?,
    val gasScheduleId: String,
    val metadataUriCid: String?,
    val vkBytesCid: String?,
    val activationHeight: Long?,
    val withdrawHeight: Long?,
    val status: VerifyingKeyStatus,
) {

    private val _inlineKeyBytes: ByteArray? = inlineKeyBytes?.copyOf()

    val inlineKeyBytes: ByteArray? get() = _inlineKeyBytes?.copyOf()

    /**
     * Serialises the record into Norito-style arguments using the provided backend identifier for
     * inline verifying key bytes.
     */
    fun toArguments(backend: String): Map<String, String> {
        val args = linkedMapOf<String, String>()
        args["record.version"] = version.toUInt().toString()
        args["record.circuit_id"] = circuitId
        args["record.backend_tag"] = backendTag.noritoValue
        args["record.curve"] = curve
        args["record.public_inputs_schema_hash_hex"] = schemaHashHex
        args["record.commitment_hex"] = commitmentHex
        if (_inlineKeyBytes != null) {
            args["record.vk_bytes_b64"] = Base64.encode(_inlineKeyBytes)
            args["record.vk_len"] = vkLength!!.toUInt().toString()
        } else if (vkLength != null) {
            args["record.vk_len"] = vkLength.toUInt().toString()
        }
        if (maxProofBytes != null) {
            args["record.max_proof_bytes"] = maxProofBytes.toUInt().toString()
        }
        args["record.gas_schedule_id"] = gasScheduleId
        if (metadataUriCid != null) {
            args["record.metadata_uri_cid"] = metadataUriCid
        }
        if (vkBytesCid != null) {
            args["record.vk_bytes_cid"] = vkBytesCid
        }
        if (activationHeight != null) {
            args["record.activation_height"] = activationHeight.toULong().toString()
        }
        if (withdrawHeight != null) {
            args["record.withdraw_height"] = withdrawHeight.toULong().toString()
        }
        args["record.status"] = status.wireName
        return args
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is VerifyingKeyRecordDescription) return false
        return version == other.version
            && circuitId == other.circuitId
            && backendTag == other.backendTag
            && curve == other.curve
            && schemaHashHex == other.schemaHashHex
            && commitmentHex == other.commitmentHex
            && _inlineKeyBytes.contentEquals(other._inlineKeyBytes)
            && vkLength == other.vkLength
            && maxProofBytes == other.maxProofBytes
            && gasScheduleId == other.gasScheduleId
            && metadataUriCid == other.metadataUriCid
            && vkBytesCid == other.vkBytesCid
            && activationHeight == other.activationHeight
            && withdrawHeight == other.withdrawHeight
            && status == other.status
    }

    override fun hashCode(): Int {
        var result = version.hashCode()
        result = 31 * result + circuitId.hashCode()
        result = 31 * result + backendTag.hashCode()
        result = 31 * result + curve.hashCode()
        result = 31 * result + schemaHashHex.hashCode()
        result = 31 * result + commitmentHex.hashCode()
        result = 31 * result + _inlineKeyBytes.contentHashCode()
        result = 31 * result + (vkLength?.hashCode() ?: 0)
        result = 31 * result + (maxProofBytes?.hashCode() ?: 0)
        result = 31 * result + gasScheduleId.hashCode()
        result = 31 * result + (metadataUriCid?.hashCode() ?: 0)
        result = 31 * result + (vkBytesCid?.hashCode() ?: 0)
        result = 31 * result + (activationHeight?.hashCode() ?: 0)
        result = 31 * result + (withdrawHeight?.hashCode() ?: 0)
        result = 31 * result + status.hashCode()
        return result
    }

    companion object {

        /**
         * Creates a validated `VerifyingKeyRecordDescription`.
         *
         * Either `inlineKeyBytes` must be provided (and the commitment is computed from `backend`
         * + bytes), or both `commitmentHex` and `vkLength` must be supplied.
         *
         * @param backend backend identifier used to compute the commitment hash for inline keys
         */
        fun create(
            backend: String,
            version: Int,
            circuitId: String,
            schemaHashHex: String,
            gasScheduleId: String,
            backendTag: VerifyingKeyBackendTag = VerifyingKeyBackendTag.UNSUPPORTED,
            curve: String = "unknown",
            commitmentHex: String? = null,
            inlineKeyBytes: ByteArray? = null,
            vkLength: Int? = null,
            maxProofBytes: Int? = null,
            metadataUriCid: String? = null,
            vkBytesCid: String? = null,
            activationHeight: Long? = null,
            withdrawHeight: Long? = null,
            status: VerifyingKeyStatus = VerifyingKeyStatus.ACTIVE,
        ): VerifyingKeyRecordDescription {
            require(version >= 0) { "version must be non-negative" }

            val trimmedCircuitId = circuitId.trim()
            require(trimmedCircuitId.isNotEmpty()) { "circuitId must not be blank" }

            val normalizedSchemaHash = validateHex(schemaHashHex, "public_inputs_schema_hash_hex")
            requireNotNull(normalizedSchemaHash) { "public_inputs_schema_hash_hex must be provided" }

            val trimmedGasScheduleId = gasScheduleId.trim()
            require(trimmedGasScheduleId.isNotEmpty()) { "gasScheduleId must not be blank" }

            if (vkLength != null) require(vkLength > 0) { "vkLength must be greater than zero" }
            if (maxProofBytes != null) require(maxProofBytes >= 0) { "maxProofBytes must be non-negative" }
            if (activationHeight != null) require(activationHeight >= 0) { "activationHeight must be non-negative" }
            if (withdrawHeight != null) require(withdrawHeight >= 0) { "withdrawHeight must be non-negative" }

            val normalizedCommitmentHex = validateHex(commitmentHex, "commitment_hex")
            val normalizedCurve = curve.trim().ifEmpty { "unknown" }
            val normalizedMetadataUriCid = metadataUriCid?.trim()?.ifEmpty { null }
            val normalizedVkBytesCid = vkBytesCid?.trim()?.ifEmpty { null }

            val inlineBytes = inlineKeyBytes?.copyOf()
            val resolvedVkLength: Int?
            if (inlineBytes != null) {
                resolvedVkLength = inlineBytes.size
                if (vkLength != null) {
                    check(vkLength == resolvedVkLength) {
                        "vkLength does not match length of inline key bytes"
                    }
                }
            } else {
                checkNotNull(normalizedCommitmentHex) {
                    "commitmentHex must be provided when vk bytes are absent"
                }
                checkNotNull(vkLength) {
                    "vkLength must be provided when vk bytes are absent"
                }
                resolvedVkLength = vkLength
            }

            if (withdrawHeight != null && activationHeight != null) {
                check(withdrawHeight >= activationHeight) {
                    "withdrawHeight must be >= activationHeight"
                }
            }

            val computedCommitment = if (inlineBytes != null) {
                computeCommitmentHex(backend, inlineBytes)
            } else {
                normalizedCommitmentHex!!.lowercase()
            }

            if (normalizedCommitmentHex != null && inlineBytes != null) {
                check(normalizedCommitmentHex.equals(computedCommitment, ignoreCase = true)) {
                    "commitmentHex does not match computed inline key commitment"
                }
            }

            return VerifyingKeyRecordDescription(
                version = version,
                circuitId = trimmedCircuitId,
                backendTag = backendTag,
                curve = normalizedCurve,
                schemaHashHex = normalizedSchemaHash,
                commitmentHex = computedCommitment,
                inlineKeyBytes = inlineBytes,
                vkLength = resolvedVkLength,
                maxProofBytes = maxProofBytes,
                gasScheduleId = trimmedGasScheduleId,
                metadataUriCid = normalizedMetadataUriCid,
                vkBytesCid = normalizedVkBytesCid,
                activationHeight = activationHeight,
                withdrawHeight = withdrawHeight,
                status = status,
            )
        }

        private fun validateHex(value: String?, field: String): String? {
            if (value == null) return null
            val normalized = value.trim()
            if (normalized.isEmpty()) return null
            require(normalized.length == 64) {
                "$field must contain exactly 64 hexadecimal characters"
            }
            for (c in normalized) {
                require(c in '0'..'9' || c in 'a'..'f' || c in 'A'..'F') {
                    "$field must contain only hexadecimal characters"
                }
            }
            return normalized.lowercase()
        }

        private fun computeCommitmentHex(backend: String, bytes: ByteArray): String {
            val digest = MessageDigest.getInstance("SHA-256")
            digest.update(backend.toByteArray())
            digest.update(bytes)
            return digest.digest().joinToString("") { "%02x".format(it) }
        }
    }
}
